ops/internal/manage/generate.go

```diff
@@ -32,21 +51,55 @@ func GenerateChainArtifacts(st state.State, wd string, shortName string, name *s
 	}
 
 	output.WriteOK("reading genesis")
-	genesis, _, err := inspect.GenesisAndRollup(&st, st.AppliedIntent.Chains[idx].ID)
+	opaqueGenesis, err := opd.InspectGenesis(statePath, strconv.FormatUint(cfg.ChainID, 10))
 	if err != nil {
-		return fmt.Errorf("failed to get genesis %d of %d: %w", idx, len(st.AppliedIntent.Chains), err)
+		return fmt.Errorf("failed to get genesis: %w", err)
 	}
 
 	stagingDir := paths.StagingDir(wd)
 
 	output.WriteOK("writing chain config")
-	if err := paths.WriteTOMLFile(path.Join(stagingDir, cfg.ShortName+".toml"), cfg); err != nil {
-		return fmt.Errorf("failed to write chain config %d of %d: %w", idx, len(st.AppliedIntent.Chains), err)
+	if err := paths.WriteTOMLFile(path.Join(stagingDir, shortName+".toml"), cfg); err != nil {
+		return fmt.Errorf("failed to write chain config at index %d: %w", idx, err)
 	}
 
+	// TODO: determine if the genesis is deterministic through these conversions
 	output.WriteOK("writing genesis")
-	if err := WriteGenesis(wd, path.Join(stagingDir, cfg.ShortName+".json.zst"), genesis); err != nil {
-		return fmt.Errorf("failed to write genesis %d of %d: %w", idx, len(st.AppliedIntent.Chains), err)
+	genesis, err := opaqueToGenesis(opaqueGenesis)
+	if err != nil {
+		return fmt.Errorf("failed to convert opaque genesis to core.Genesis: %w", err)
+	}
+
+	if err := WriteGenesis(wd, path.Join(stagingDir, shortName+".json.zst"), genesis); err != nil {
+		return fmt.Errorf("failed to write genesis at index %d: %w", idx, err)
+	}
+
+	// Copy state file to staging directory
+	output.WriteOK("writing state.json")
+	stateData, err := os.ReadFile(statePath)
+	if err != nil {
+		return fmt.Errorf("failed to read state file: %w", err)
+	}
+
+	if err := os.WriteFile(path.Join(stagingDir, "state.json"), stateData, 0o644); err != nil {
+		return fmt.Errorf("failed to write state file to staging directory: %w", err)
 	}
 	return nil
 }
+
+// Convert OpaqueMapping to core.Genesis
+func opaqueToGenesis(opaque *deployer.OpaqueMap) (*core.Genesis, error) {
+	// Step 1: Marshal the OpaqueMapping to JSON
+	jsonData, err := json.Marshal(opaque)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal OpaqueMapping to JSON: %w", err)
+	}
+
+	// Step 2: Unmarshal the JSON data into a core.Genesis struct
+	var genesis core.Genesis
+	if err := json.Unmarshal(jsonData, &genesis); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal JSON to Genesis: %w", err)
+	}
+
+	return &genesis, nil
+}
```

Comment:
This is not always safe: it is possible to get an `opaqueMap` from a _future_ version of op-deployer (say with a new hardfork time in it), and then we would lose information when converting to the `core.Genesis` if we are pinned to an older version of op-geth in the superchain registry.

If possible, we should work entirely with `opaqueMap`.

Reply:
Added the logic to do this, plus a test.

Reply:
we should always use opaquemap